tests/utilities/fault_tolerant.py

```diff
@@ -0,0 +1,101 @@
+# limitations under the License.
+
+"""
+Note: This file is used to ensure Fault Tolerant is working properly when
+Lightning is used in an environment where users have full control on their
+Kubernetes Environnement
+
+Here are the steps to use this file to ensure your cluster is properly setup for Fault Tolerant Training.
+
+1. Launch this script with `python tests/utilities/fault_tolerant.py`
+2. Once ``READY TO BE KILLED WITH SIGTERM SIGNAL`` is detected within
+   the Pod Logs, send a SIGTERM to this process. The process is waiting for it.
+3. You should detect `.pl_auto_save.ckpt exists` within the Pod Logs.
+4. Resume your instance with the same volumes and relaunch the same script.
+5. The script should contain `[-1.0939, -0.4306]` within its Pod Logs.
+
+The mode weights with the failure matches the weight without.
+The training has been properly resumed and is fully reproduced.
+
+"""
+
+# Note, this file is used to ensure Fault Tolerant is working as expected
+import os
+from time import sleep
+
+import torch
+
+from pytorch_lightning import seed_everything
+from tests.utilities.test_auto_restart import _run_training, CustomException, RandomGetItemDataset, TestModel
+
+
+class SignalTestModel(TestModel):
+    def training_step(self, batch, batch_idx):
+        if self.global_step == self.fail_on_step:
+            print("READY TO BE KILLED WITH SIGTERM SIGNAL.")
+            while not self.trainer._terminate_gracefully:
+                sleep(0.00001)
+            raise CustomException()
+        batch = batch["data"] if isinstance(batch, dict) else batch
+        self.seen_batches.append(torch.stack(batch) if isinstance(batch, list) else batch)
+        loss = sum(self.layer(b).sum() for b in batch)
+        return loss
+
+
+tmpdir = "/tmp/pl_fault_tolerant"
+
+os.makedirs(tmpdir, exist_ok=True)
+
+env_backup = os.environ.copy()
+
+auto_restart_checkpoint_path = os.path.join(tmpdir, ".pl_auto_save.ckpt")
+auto_restart_checkpoint_path_exists = os.path.exists(auto_restart_checkpoint_path)
+
+seed_everything(42)
+
+os.environ["PL_FAULT_TOLERANT_TRAINING"] = "1"
+
+dataset_classes = [RandomGetItemDataset]
+
+trainer_kwargs = dict(
+    default_root_dir=tmpdir,
+    max_epochs=3,
+    enable_progress_bar=False,
+    enable_model_summary=False,
+)
+
+if auto_restart_checkpoint_path_exists:
+    fail_on_step = -1
+    completed_batches = 5
+else:
+    fail_on_step = 4
+    completed_batches = 4
+
+# Perform a failure
+complete_batches, weights = _run_training(
+    trainer_kwargs, dataset_classes, fail_on_step=fail_on_step, model_cls=SignalTestModel
+)
+assert len(complete_batches) == completed_batches
+
+if not auto_restart_checkpoint_path_exists:
+    checkpoint_path = os.path.join(tmpdir, ".pl_auto_save.ckpt")
+    assert os.path.exists(checkpoint_path)
+    print(".pl_auto_save.ckpt exists.")
+
+if auto_restart_checkpoint_path_exists:
+    print([w for w in weights])
+
+os.environ.clear()
+os.environ.update(env_backup)
```

Comment:
this code all runs on import. this is probably not intentional

Reply:
This is intended. This code is used to validate Fault Tolerant Training works and I was thinking it could be exposed and used by organisation to validate their auto restart is working as expected.

Reply:
But the docs you added say it is meant to be run as a script.

Reply:
I have updated the docs.

Reply:
I insist. I see no other way. As you can see, the test suite never starts, because it hangs at collecting the tests. You have this pice of code:
```py
 log.info("READY TO BE KILLED WITH SIGTERM SIGNAL.")
            while not self.trainer._terminate_gracefully:
                sleep(0.1)
```
which stalls execution indefinitely.
This is because the code runs at import!! I cannot see how this is intentional. Please consider adding if `if __name__ == "__main__"` check or let me know why this is not possible.

Reply:
Done !